src/pydicom/filewriter.py

```diff
@@ -625,7 +646,19 @@ def write_data_element(
         # unless is SQ with undefined length.
         fp.write_UL(0xFFFFFFFF if is_undefined_length else value_length)
 
-    fp.write(buffer.getvalue())
+    # if the element was buffered, we wrote directly to the file already
+    if elem.is_buffered:
+        # write to the file directly
+        fn(fp, elem)
+    else:
+        # copy the buffer to the file
+        fp.write(buffer.getvalue())
+        # write the length now that we know it
+        curr_pos = fp.tell()
+        fp.seek(vr_length_seek_pos)
+        fp.write_UL(elem.bytes_read_from_buffer)
+        fp.seek(curr_pos)
```

Comment:
This feels pretty hacky - I'd love to get some thoughts on how to handle this. I'm stuck on how to handle writing the length of data before we know the length of data without having to backtrack without calling `stat` on the original file the buffer is from.

Reply:
I'm not sure I follow the logic here, but I see no reason why you have to first write the buffer before writing the length, as you can get the length from the buffer. Also note that we don't want to seek backwards at all as this breaks certain use cases where fp is a stream or a zip file I think (I don't remember what use case it was exactly).

Reply:
> I'm not sure I follow the logic here, but I see no reason why you have to first write the buffer before writing the length, as you can get the length from the buffer

The problem is that I'm not writing the data to the buffer but instead to the file directly. The goal is to avoid reading the entire buffered element into memory and instead to transfer the data directly from the element (which contains a buffered reader) to the dicom file.

For a little context, the goal is to avoid reading large uncompressed files into memory.

Reply:
> Edit: Ok, the problem was with writing gzip file-like (see https://github.com/pydicom/pydicom/issues/753). There was also an old issue about writing to a stream (https://github.com/pydicom/pydicom/issues/154).

That's quite the conundrum... any thoughts on how to tackle it then without reading the entirety of PixelData into a memory buffer?

Reply:
> The problem is that I'm not writing the data to the buffer but instead to the file directly.

A right, I should have seen that - thanks!

> That's quite the conundrum... any thoughts on how to tackle it then without reading the entirety of PixelData into a memory buffer?

Right now I have no idea. May have to think about it... How easy would it be to configure that no buffering is used? That would still allow the use case where seeking back is not possible (in case we find no other solution).